docs/src/architecture/notes.md

```diff
@@ -1,73 +1,191 @@
+  
+  If execution hasn't failed, the script invokes a helper procedure `exec.add_note_assets_to_account` to add the note's assets into the executing account's vault.
+
+  ### Add assets
+  This procedure adds the assets held by the note into the account's vault. 
+
+  ```
+  #! Helper procedure to add all assets of a note to an account.
+  #!
+  #! Inputs: []
+  #! Outputs: []
+  #!
+  proc.add_note_assets_to_account
+      push.0 exec.note::get_assets
+      # => [num_of_assets, 0 = ptr, ...]
+  
+      # compute the pointer at which we should stop iterating
+      dup.1 add
+      # => [end_ptr, ptr, ...]
+  
+      # pad the stack and move the pointer to the top
+      padw movup.5
+      # => [ptr, 0, 0, 0, 0, end_ptr, ...]
+  
+      # compute the loop latch
+      dup dup.6 neq
+      # => [latch, ptr, 0, 0, 0, 0, end_ptr, ...]
+  
+      while.true
+          # => [ptr, 0, 0, 0, 0, end_ptr, ...]
+  
+          # save the pointer so that we can use it later
+          dup movdn.5
+          # => [ptr, 0, 0, 0, 0, ptr, end_ptr, ...]
+  
+          # load the asset and add it to the account
+          mem_loadw call.wallet::receive_asset
+          # => [ASSET, ptr, end_ptr, ...]
+  
+          # increment the pointer and compare it to the end_ptr
+          movup.4 add.1 dup dup.6 neq
+          # => [latch, ptr+1, ASSET, end_ptr, ...]
+      end
+  
+      # clear the stack
+      drop dropw drop
+  end
+  ```
+
+  The procedure starts by calling `exec.note::get_assets` and putting the note's number of assets and the memory pointer of the first asset on top of the stack. Assets are stored in consecutive memory slots, so `dup.1 add` provides the last memory slot. Because [assets](assets.md) are represented by `Words` in Miden Assembly, the procedure pads the stack with four `0`s. Now, if there is at least one asset (checked by `dup dup.6 neq`), the loop starts. It first saves the pointer for later use (`dup movdn.5`), then loads the first asset `mem_loadw` on top of the stack. Now, the procedure calls the a function of the account interface `call.wallet::receive_asset` to put the asset into the account's vault. The note script cannot directly call an account function to add the asset. The account must expose this function in its interface. Lastly, the pointer gets incremented, and if there is a second asset, the loop continues (`movup.4 add.1 dup dup.6 neq`). Finally, when all assets were put into the account's vault, the stack is cleared (`drop dropw drop`).
```

Comment:
This is a pretty long paragraph which packs a lot of content. I would probably split this up.

Also, we should probably explain more about the wallet interface somewhere (or maybe we already do that?) and have a link here to that section. This will make it clear that there are also standardized account interfaces (similar to standardized notes). This can be done in a different PR.